modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java

```diff
@@ -143,14 +145,27 @@ private CompletableFuture<?> onIndexBuilding(StartBuildingIndexEventParameters p
 
             for (TablePartitionId primaryReplicaId : primaryReplicaIds) {
                 if (primaryReplicaId.tableId() == indexDescriptor.tableId()) {
-                    CompletableFuture<?> startBuildIndexFuture = getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId)
-                            .thenCompose(mvTableStorage -> awaitPrimaryReplica(primaryReplicaId, clockService.now())
-                                    .thenAccept(replicaMeta -> tryScheduleBuildIndex(
-                                            primaryReplicaId,
-                                            indexDescriptor,
-                                            mvTableStorage,
-                                            replicaMeta
-                                    ))
+                    int tableId = indexDescriptor.tableId();
+
+                    CompletableFuture<?> startBuildIndexFuture = getMvTableStorageFuture(parameters.causalityToken(), tableId)
+                            .thenCompose(mvTableStorage -> {
+                                if (mvTableStorage == null) {
+                                    throw new IgniteInternalException(
```

Comment:
I think it will be enough for us to add a `requireMvTableStorageNonNull `and throw an exception with the general text; the stack trace will make it clear where the error occurred.

You could also keep the existing code and add something like:
```
    private CompletableFuture<MvTableStorage> getMvTableStorageFuture(long causalityToken, TablePartitionId replicaId) {
        return indexManager.getMvTableStorage(causalityToken, replicaId.tableId())
                .thenApply(IndexBuildController::requireMvTableStorageNonNull);
    }

    private static MvTableStorage requireMvTableStorageNonNull(MvTableStorage mvTableStorage) {
        ...
    }
```

Reply:
good proposal, thanks!